docs/tech-notes/life_of_a_query.md

```diff
@@ -1154,237 +1162,202 @@ and the latest version visible to a given transaction is given by its
-The command has now been applied (on one particular replica, but keep in mind
-that the process described in this section happens on *every* replica).
+The command has now been applied on this particular replica, but keep in
+mind that this process happens on _every_ replica.
 
-We've glossed over something important: after applying the command, on the
-replica where the client is waiting for the response (i.e. typically the lease
-holder), we need to signal the channel on which the client is waiting (which, as
-we saw in the previous section, is blocked in `executeWriteBatch`). This happens
+Let's not forget about the waiting proposer, who is still blocked on the
+proposal channel back in `executeWriteBatch()`. After applying the command, on
+the proposing replica (i.e. typically the lease holder), we need to return the
+result back through the channel. This happens
 [at the end](https://github.com/cockroachdb/cockroach/blob/6e4c3ae05d92f4104b7b9fbb33f9d1cecf055fd3/pkg/kv/kvserver/apply/task.go#L297)
-of command application, where [`AckOutcomeAndFinish`](https://github.com/cockroachdb/cockroach/blob/6e4c3ae05d92f4104b7b9fbb33f9d1cecf055fd3/pkg/kv/kvserver/replica_application_cmd.go#L166)
-is called on every command -- if the command was local, it will eventually
+of command application, where [`AckOutcomeAndFinish()`](https://github.com/cockroachdb/cockroach/blob/6e4c3ae05d92f4104b7b9fbb33f9d1cecf055fd3/pkg/kv/kvserver/replica_application_cmd.go#L166)
+gets called on every command: if the command was local, it will eventually
 [signal](https://github.com/cockroachdb/cockroach/blob/e4bb550d9abae97ab09c099592f32f88ad895177/pkg/kv/kvserver/replica_proposal.go#L145)
-a proposal result by [sending it](https://github.com/cockroachdb/cockroach/blob/e4bb550d9abae97ab09c099592f32f88ad895177/pkg/kv/kvserver/replica_proposal.go#L161)
-through the channel that the proposer is waiting on.
-
-We've now come full circle - the proposer will now be unblocked and receive a
-response on the channel, and it can unwind the stack letting its client know
-that the request is complete. This reply can travel through the hierarchy of
-`Sender`s, back from the lease holder node to the SQL gateway node, to a SQL
-tree of `planNode`s, to the SQL Executor, and, through the `pgwire`
-implementation, to the SQL client.
+the proposer by [sending](https://github.com/cockroachdb/cockroach/blob/e4bb550d9abae97ab09c099592f32f88ad895177/pkg/kv/kvserver/replica_proposal.go#L161)
+the result.
+
+We've now come full circle: the proposer unblocks and receives its result,
+unwinding the stack to let the client know that the request is complete. The
+response travels back through the hierarchy of `Sender`s, from the lease holder
+node to the gateway node, through the DistSQL flow and transaction coordinator,
+via the session manager and `pgwire` implementation, before finally reaching the
+SQL client.
```

Comment:
This isn't how it works most of the time, due to write pipelining (though it does happen that way for the STAGING transaction record). For the other writes the proposer just waits for the evaluation, and then returns, verifying that the writes happened later. This is probably not worth getting into here though.

Reply:
Yeah, I haven't wanted to get into the transactional model here -- that's more of a multi-query concept and we're mostly concerned with tracing a single query here. Not to mention that it'd fill an entire article on its own.

However, maybe it's worth just mentioning it in passing, since it's fairly important. We reference it very briefly in the `TxnCoordSender` section, let me try to add a short paragraph on this there.

Reply:
Added a paragraph at the end of "Request Evaluation", seemed more appropriate there since the only effect is basically that we return immediately without waiting for application.